### PR TITLE
make a copy of file before upload on ComposeActivity => solve upload failed bug

### PR DESCRIPTION
**Pull request by lenchan139**
Thanks @charlag , I solved it by make a copy to cache before upload :)

~~Hi, I bought Samsung S10e currently and try to use Tusky on it. But there have a problem if I upload photos that taken by stock camera. To fix it, the simplest way is always to compress photo before uploading to mastodon instance.~~

### CODE REVIEW

**Comment by charlag** [comment]
Hey
Can you please explain what is the problem with uploads? I think we would prefer to make solution which works automatically rather than adding an option

**Comment by lenchan139** [comment]
Hi @charlag ,
Yeah, You are right. :(
###
If I upload a photo and it was taken by the Samsung stock camera, there have some photo will cause the error from OkHttp and it says:
`java.net.ProtocolException: expected 4021 bytes but received 4182`

But, if I upload a re-save version(no matter upload to telegram and download it, use Snapseed re-save it) with that photo and upload it, it will upload success.

**Comment by charlag** [comment]
Hm, thanks.
Doe sit happen with every app which provides media?

**Comment by lenchan139** [comment]
No.
I can upload same photo on another Mastodon client called Fedilab and It happens only Tusky. :(

**Comment by charlag** [comment]
then we better find the error :thinking: 
we can try to copy it into our own storage first and then upload, not sure if it would help

**Comment by lenchan139** [comment]
Oh, many thank @charlag !!!
It works now without error if I make a copy to cache and upload the cached copy.

**Comment by charlag** [comment]
Hey, thank you, I encountered similar thing in other apps, we'll try to review it in reasonable time 😅

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 1]
```diff
@@ -1431,13 +1432,36 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
     }
 
 
-    private void pickMedia(Uri uri, long mediaSize, String description) {
+    private void pickMedia(Uri inUri, long mediaSize, String description) {
+        Uri uri = inUri;
+        ContentResolver contentResolver = getContentResolver();
+        String mimeType = contentResolver.getType(uri);
+        String surfix = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType);
```
Comment (Tak) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: Why not use the filename suffix from the uri?

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 2]
```diff
@@ -1431,13 +1432,36 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
...
+        try {
+            InputStream tempInput;
+            FileOutputStream out;
```
Comment (Tak) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: When do these get closed?

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 3]
```diff
@@ -1431,13 +1432,36 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
...
+            }
+            uri = FileProvider.getUriForFile(this,
+                    BuildConfig.APPLICATION_ID+".fileprovider",
+                    file);
+            Log.v("checker","a");
+        }catch(IOException e){
+            e.printStackTrace();
+            Log.w(TAG, e);
+            uri = inUri;
+        }
```
Comment (Tak) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: Does `mediaSize` need to be updated?

**Comment by charlag** [comment]
@lenchan139 hi! Do you plan to continue working on this?

**Comment by lenchan139** [comment]
@charlag Hi, It is done now. :)

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 4]
```diff
@@ -1431,13 +1432,42 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
...
+            }
+            uri = FileProvider.getUriForFile(this,
+                    BuildConfig.APPLICATION_ID+".fileprovider",
+                    file);
+
+            mediaSize = getMediaSize(getContentResolver(), uri);
+            tempInput.close();
+            out.close();
+        }catch(IOException e){
+            e.printStackTrace();
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: I think we should try to close streams on error?

File: app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java [hunk 5]
```diff
@@ -1431,13 +1432,42 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
...
+            Log.w(TAG, e);
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java]: `e.printStackTrace();` is deprecated, `Log.w(TAG, e);` alone should be fine.

**Comment by connyduck** [comment]
there is a `closeQuietly` method in IOUtils you can use to make this way shorter

**Comment by lenchan139** [comment]
repoen at #1294

**Pull request closed without merging**